File: Data Structures & Algorithms/combination-target-sum/submission-16.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) { 

        let result = [] 

        function dfs(j, cur, curSum) {
            if (j >= nums.length || curSum > target) {
                return
            } 

            if (curSum === target) {
                result.push([...cur]) 
                return
            } 

            for (let i = j; i < nums.length; i ++) {
                cur.push(nums[i]) 
                dfs(i, cur, curSum + nums[i]) 
                cur.pop()
            }
        } 

        dfs(0, [], 0) 

        return result
    }
}

File: Data Structures & Algorithms/count-paths/submission-16.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        def dfs(m, n): 
            if m == 1 or n == 1: 
                return 1 

            return dfs(m - 1, n) + dfs(m, n - 1) 

        return dfs(m, n)